File: backend/Services/ProjectService.cs
using backend.Models.Projects;
using backend.Repositories;

namespace backend.Services
{
    public class ProjectService : IProjectService
    {
        private IProjectRepository ProjectRepository;
        public ProjectService(IProjectRepository projectRepository)
        {
            ProjectRepository = projectRepository;
        }
        public async Task<IEnumerable<Project>> GetAllProjectsAsync(int userId)
        {
            return await ProjectRepository.GetAllProjectsAsync(userId);
        }

        public async Task<Project?> GetProjectDetailsAsync(int projectId)
        {
            return await ProjectRepository.GetProjectDetailsAsync(projectId);
        }

        public async Task<IEnumerable<ProjectUserMapping>> GetProjectUserMappingsAsync()
        {
            return await ProjectRepository.GetProjectUserMappingsAsync();
        }

        public async Task<int> UpsertProjectAsync(ProjectDetails project)
        {
            return await ProjectRepository.UpsertProjectAsync(project);
        }
    }
}

File: frontend/src/app/services/project.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Project, ProjectDetails, ProjectUserMapping } from '../models/project';

@Injectable({
  providedIn: 'root',
})
export class ProjectService {
  constructor(private httpClient: HttpClient) {}

  private baseUrl = 'https://localhost:5001';

  GetProjects(userId: number): Observable<Project[]> {
    return this.httpClient.get<Project[]>(
      `${this.baseUrl}/projects/users/${userId}`
    );
  }

  GetProject(projectId: number): Observable<Project> {
    return this.httpClient.get<Project>(
      `${this.baseUrl}/projects/${projectId}`
    );
  }

  GetProjectUserMappings(): Observable<ProjectUserMapping[]> {
    return this.httpClient.get<ProjectUserMapping[]>(
      `${this.baseUrl}/projects/users`
    );
  }

  UpsertProject(project: ProjectDetails): Observable<number> {
    return this.httpClient.put<number>(`${this.baseUrl}/projects`, project);
  }
}
